From b0c80a72a9a32e17953483e1fb6b66dcfd8d0096 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=D0=BD=D0=B0=D0=B1?= <nabijaczleweli@nabijaczleweli.xyz>
Date: Tue, 12 Jan 2021 18:28:26 +0100
Subject: [PATCH] Fall back from link(2) to rename(2) if the underlying
 filesystem does not support it
X-Mutt-PGP: OS

Most notably, this likely fixes kernel upgrades with /boot on FAT
---
 src/archives.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/src/archives.c b/src/archives.c
index 96079ab1f..f38deee0a 100644
--- a/src/archives.c
+++ b/src/archives.c
@@ -1061,9 +1061,16 @@ tarobject(struct tar_archive *tar, struct tar_entry *ti)
       tarobject_set_se_context(fnamevb.buf, fnametmpvb.buf, stab.st_mode);
     } else {
       debug(dbg_eachfiledetail, "tarobject nondirectory, 'link' backup");
-      if (link(fnamevb.buf,fnametmpvb.buf))
-        ohshite(_("unable to make backup link of '%.255s' before installing new version"),
-                ti->name);
+      if (link(fnamevb.buf,fnametmpvb.buf)) {
+        if (errno == EPERM) {
+          debug(dbg_eachfiledetail, "unable to link, falling back to nonatomic");
+          if (rename(fnamevb.buf,fnametmpvb.buf))
+            ohshite(_("unable to move aside '%.255s' to install new version"),
+                    ti->name);
+        } else
+          ohshite(_("unable to make backup link of '%.255s' before installing new version"),
+                  ti->name);
+      }
     }
   }
 
-- 
2.20.1

